src/libp2p/peers.rs

```diff
@@ -328,7 +323,12 @@ where
                                 (actual_peer_index, collection::ConnectionId::min_value())
                                     ..=(actual_peer_index, collection::ConnectionId::max_value()),
                             )
-                            .filter(|(_, established)| **established)
+                            .filter(|(_, connection_id)| {
+                                // Note that connections that are shutting down are still counted,
+                                // as we report the disconnected event only at the end of the
+                                // shutdown.
+                                self.inner.connection_state(*connection_id).established
```

Comment:
is it possible that this code panics when connection existed during range, but was shutdown before filter => panic?

Reply:
No, the local list of connections is always strictly in sync with the underlying layer.